eli5/keras/explain_prediction.py

```diff
@@ -253,90 +451,253 @@ def _extract_image(doc):
     elif isinstance(layer, int):
-        activation_layer = model.get_layer(index=layer)
+        # keras.get_layer() performs a bottom-up horizontal graph traversal
+        # the function raises ValueError if the layer index / name specified is not found
+        return model.get_layer(index=layer)
     elif isinstance(layer, str):
-        activation_layer = model.get_layer(name=layer)
+        return model.get_layer(name=layer)
     else:
-        raise TypeError('Invalid layer (must be str, int, keras.layers.Layer, or None): %s' % layer)
+        raise TypeError('Invalid layer (must be str, int, or keras.layers.Layer): %s' % layer)
 
-    if _is_suitable_activation_layer(model, activation_layer):
-        # final validation step
-        return activation_layer
-    else:
-        raise ValueError('Can not perform Grad-CAM on the retrieved activation layer')
 
+# Heuristics for getting a suitable activation layer
 
-def _search_layer_backwards(model, condition):
-    # type: (Model, Callable[[Model, Layer], bool]) -> Layer
-    """
-    Search for a layer in ``model``, backwards (starting from the output layer),
-    checking if the layer is suitable with the callable ``condition``,
-    """
-    # linear search in reverse through the flattened layers
-    for layer in model.layers[::-1]:
-        if condition(model, layer):
-            # linear search succeeded
-            return layer
-    # linear search ended with no results
-    raise ValueError('Could not find a suitable target layer automatically.')        
+def _autoget_layer_image(model):
+    # type: (Model) -> Layer
+    """Try find a suitable layer for image ``model``."""
+    l = _search_layer(model, _backward_layers, _is_suitable_image_layer)
+    return l if l is not None else _middle_layer(model)
 
 
-def _is_suitable_activation_layer(model, layer):
+def _is_suitable_image_layer(model, layer):
     # type: (Model, Layer) -> bool
-    """
-    Check whether the layer ``layer`` matches what is required 
-    by ``model`` to do Grad-CAM on ``layer``.
-    Returns a boolean.
-
-    Matching Criteria:
-        * Rank of the layer's output tensor.
+    """Check whether the layer ``layer`` matches what is required
+    by ``model`` to do Grad-CAM on ``layer``, for image-based models.
     """
     # TODO: experiment with this, using many models and images, to find what works best
-    # Some ideas: 
+    # Some ideas:
     # check layer type, i.e.: isinstance(l, keras.layers.Conv2D)
     # check layer name
+    # input wrpt output
 
     # a check that asks "can we resize this activation layer over the image?"
     rank = len(layer.output_shape)
     required_rank = len(model.input_shape)
     return rank == required_rank
+
+
+def _autoget_layer_text(model, character=False):
+    # type: (Model, Union[np.ndarray, list]) -> Layer
+    """Try find a suitable layer for text ``model``.
+    If ``character`` is `True`, tokenization is character-level.
+    """
+    if character:
```

Comment:
As we discussed, let's try to return the last spatial layer in all cases instead?